src/raw_node.rs

```diff
@@ -349,94 +384,287 @@ impl<T: Storage> RawNode<T> {
         Err(Error::StepPeerNotFound)
     }
 
-    /// Given an index, creates a new Ready value from that index.
-    pub fn ready_since(&mut self, applied_idx: u64) -> Ready {
-        Ready::new(
-            &mut self.raft,
-            &self.prev_ss,
-            &self.prev_hs,
-            Some(applied_idx),
-        )
-    }
-
     /// Ready returns the current point-in-time state of this RawNode.
     pub fn ready(&mut self) -> Ready {
-        Ready::new(&mut self.raft, &self.prev_ss, &self.prev_hs, None)
+        let raft = &mut self.raft;
+
+        self.max_number += 1;
+        let mut rd = Ready {
+            number: self.max_number,
+            ..Default::default()
+        };
+        let mut rd_record = ReadyRecord {
+            number: self.max_number,
+            ..Default::default()
+        };
+
+        // If there is a pending snapshot, do not get the whole Ready.
+        if self.pending_snapshot {
+            self.records.push_back(rd_record);
+            return rd;
+        }
+
+        if self.prev_ss.raft_state != StateRole::Leader && raft.state == StateRole::Leader {
+            // The vote msg which makes this peer become leader has been sent after persisting.
+            // So the remain records must be generated during being candidate which can not
+            // have last_log and snapshot(if so, it should become to follower). The only things
+            // left are messages and they can be sent without persisting because no key data changes
+            // (term, vote, entry). These messages should be added before raft.msgs to avoid out of order.
+            for record in self.records.drain(..) {
+                assert_eq!(record.last_entry, None);
+                assert!(!record.has_snapshot);
+                if !record.messages.is_empty() {
+                    self.messages.push(record.messages);
+                }
+            }
+        }
```

Comment:
This may imply that we must call `commit_ready` (by functions like `advance_append` ) to update `self.prev_ss` before getting the next `Ready`, maybe we should put this constraint as a comment on `fn ready`

Reply:
Yes. The users must obey this rule. It's mentioned in the document. Let me think about the comments 🤣.